src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,169 @@
-""" This is a dummy example """
+import requests
+import pandas as pd
+import time
+import matplotlib.pyplot as plt
+from typing import Dict, Optional, List
+
+API_URL = "https://climate-api.open-meteo.com/v1/climate?"
+
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+
+VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"
+
+
+def fetch_url(url: str, params: Dict[str, str] = None) -> Optional[Dict]:
+    """
+    Fetch data from an API given a URL and optional parameters.
+    """
+    MAX_SECONDS_TO_WAIT = 1
+    MAX_ATTEMPTS = 5
+
+    for attempt in range(MAX_ATTEMPTS):
+        try:
+            response = requests.get(url, params)
+            response.raise_for_status()
+            print("Data fetched successfully!")
+            return response.json()
+
+        except requests.exceptions.RequestException as error:
+            if error.response.status_code == 404:
+                raise
+            elif error.response.status_code == 429 and attempt != MAX_ATTEMPTS - 1:
+                retry_after = int(
+                    response.headers.get("Retry-After", MAX_SECONDS_TO_WAIT)
+                )
+                time.sleep(retry_after)
+                continue
+            print(
+                f"""Data fetching unsuccessful. Error {error.response.status_code}: 
+                {error.response.reason}."""
+            )
+            return None
+
+
+def fetch_data_meteo_api(
+    latitude: float,
+    longitude: float,
+    start_date: str,
+    end_date: str,
+    climate_models: List[str] = None,
+) -> Optional[Dict]:
+    """
+    Fetch weather data from meteo API given a specified location,
+    time range in "YYYY-MM-DD" format and climate models.
+    """
+    params = {
```

Comment:
This is perfectly fine. If you wanna explore how to design data models in a principled way take a look at dataclasses. They are specially useful for handling and designing data models:

https://docs.python.org/3/library/dataclasses.html